# SQL

**Описание проекта**<br>
<br>
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. <br>
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.<br>

**Цель проекта** - проанализировать базу данных и дать ответы на поставленные вопросы. 

**Описание данных**<br>
<br>
Таблица `books`<br>
<br>
`book_id` — идентификатор книги;<br>
`author_id` — идентификатор автора;<br>
`title` — название книги;<br>
`num_pages` — количество страниц;<br>
`publication_date` — дата публикации книги;<br>
`publisher_id` — идентификатор издателя.<br>
<br>
Таблица `authors`<br>
<br>
`author_id` — идентификатор автора;<br>
`author` — имя автора.<br>
<br>
Таблица `publishers`<br>
<br>
`publisher_id` — идентификатор издательства;<br>
`publisher` — название издательства;<br>
<br>
Таблица `ratings`<br>
<br>
`rating_id` — идентификатор оценки;<br>
`book_id` — идентификатор книги;<br>
`username` — имя пользователя, оставившего оценку;<br>
`rating` — оценка книги.<br>
<br>
Таблица `reviews`<br>
<br>
`review_id` — идентификатор обзора;<br>
`book_id` — идентификатор книги;<br>
`username` — имя пользователя, написавшего обзор;<br>
`text` — текст обзора.<br>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Просмотр-таблиц" data-toc-modified-id="Просмотр-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Просмотр таблиц</a></span></li><li><span><a href="#Задания" data-toc-modified-id="Задания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задания</a></span><ul class="toc-item"><li><span><a href="#Задача-1" data-toc-modified-id="Задача-1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Задача 1</a></span></li><li><span><a href="#Задача-2" data-toc-modified-id="Задача-2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Задача 2</a></span></li><li><span><a href="#Задача-3" data-toc-modified-id="Задача-3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Задача 3</a></span></li><li><span><a href="#Задача-4" data-toc-modified-id="Задача-4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Задача 4</a></span></li><li><span><a href="#Задача-5" data-toc-modified-id="Задача-5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Задача 5</a></span></li></ul></li></ul></div>

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Просмотр таблиц

In [2]:
table = ['books', 'authors', 'ratings', 'reviews', 'publishers']

for i in table:
    query = '''
        SELECT *
        FROM {}
    '''.format(i)
    
    temp_df = pd.io.sql.read_sql(query, con = engine)
    display(temp_df.head())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

## Задания

### Задача 1

Посчитайте, сколько книг вышло после 1 января 2000 года

In [3]:
query = '''
        SELECT COUNT (*)
        FROM books
        WHERE publication_date >='2000-01-01'
    '''
task_1 = pd.io.sql.read_sql(query, con = engine)
display(task_1)

count
0    821

**Ответ:** После 01-01-2000 вышла 821 книга

### Задача 2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [4]:
query = '''
        WITH a AS 
        (SELECT b.title
            , COUNT(r.review_id) OVER (PARTITION BY r.book_id)
            , AVG(rt.rating)
        FROM books b
        left JOIN reviews r ON b.book_id = r.book_id
        left JOIN ratings rt ON b.book_id = rt.book_id
        GROUP BY 1, r.review_id
        ORDER BY 2 DESC, 3 DESC)
        SELECT *
        FROM a
        GROUP BY 1, 2, 3
        ORDER BY 2 DESC, 3 DESC
        '''
task_2 = pd.io.sql.read_sql(query, con = engine)
display(task_2)

title  count       avg
0                              Twilight (Twilight  #1)      7  3.662500
1    Harry Potter and the Prisoner of Azkaban (Harr...      6  4.414634
2    Harry Potter and the Chamber of Secrets (Harry...      6  4.287500
3                                       The Book Thief      6  4.264151
4                                     The Glass Castle      6  4.206897
..                                                 ...    ...       ...
995  Disney's Beauty and the Beast (A Little Golden...      0  4.000000
996                          Essential Tales and Poems      0  4.000000
997                               Leonardo's Notebooks      0  4.000000
998    Anne Rice's The Vampire Lestat: A Graphic Novel      0  3.666667
999                            The Natural Way to Draw      0  3.000000

[1000 rows x 3 columns]

**Ответ:** Лидером является книга `Twilight (Twilight #1)` - `Сумерки` с 7 обзорами, имеющая средную оценку 3,6625.

### Задача 3

Определите издательство, которое выпустило наибольшее число книг толще
50 страниц — так вы исключите из анализа брошюры

In [5]:
query = '''
        WITH a AS 
            (SELECT COUNT(book_id) as count
                , publisher_id
                , ROW_NUMBER() OVER (ORDER BY COUNT (book_id) DESC) AS rn
            FROM books
            GROUP BY 2)
        SELECT p.publisher
            , a.count
        FROM publishers p
        JOIN a ON a.publisher_id  = p.publisher_id
        WHERE a.rn=1
        '''
task_3 = pd.io.sql.read_sql(query, con = engine)
display(task_3)

publisher  count
0  Penguin Books     42

**Ответ:** Издательство `Penguin Books` выпустило наибольшее число книг (42 шт.) толще 50 страниц

### Задача 4

Определите автора с самой высокой средней оценкой книг — учитывайте
только книги с 50 и более оценками

In [6]:
query = '''
WITH b_rt AS (SELECT b.author_id
                , b.book_id
                , COUNT(rt.rating_id)
             FROM books b
             JOIN ratings rt ON rt.book_id = b.book_id
             WHERE b.num_pages>50
             GROUP BY 1, 2
             HAVING COUNT(rt.rating_id)>=50)
        SELECT a.author
        , AVG(rt.rating)
        FROM b_rt
        JOIN ratings rt ON rt.book_id = b_rt.book_id
        JOIN authors a ON a.author_id  = b_rt.author_id
        GROUP BY 1
        ORDER BY 2 desc
        LIMIT 1
        '''
task_4 = pd.io.sql.read_sql(query, con = engine)
display(task_4)

author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

**Ответ:** автором с самой высокой средней оценкой книг является `J.K.Rowling/Mary GrandPré` со средней оценкой 4,287097.

### Задача 5

Посчитайте среднее количество обзоров от пользователей, которые поставили
больше 50 оценок

In [7]:
query = '''
        WITH c AS 
            (SELECT rt.username
                  , COUNT(rt.rating_id) AS cnt_rt
            FROM ratings rt
            GROUP BY 1
            HAVING COUNT(rt.rating_id) > 50),
            d as
            (SELECT r.username
                  , COUNT(r.review_id) AS cnt_rew
            FROM reviews r
            GROUP BY 1)
        SELECT AVG(d.cnt_rew)
        FROM d
        JOIN c ON c.username  = d.username
        '''
task_5 = pd.io.sql.read_sql(query, con = engine)
display(task_5)

avg
0  24.333333

**Ответ:** среднее количество обзоров от пользователей составляет 24,3(3)